File: backend/app/routers/metrics.py
from fastapi import APIRouter, Query

from app.dependencies import AdminUser, DBSession
from app.schemas.metrics import (
    CompletionMetrics,
    EnrollmentsTimeSeries,
    OverviewMetrics,
    PopularCourses,
    RecentActivity,
)
from app.services import metrics_service

router = APIRouter(tags=["Admin / Metrics"])


@router.get("/overview", response_model=OverviewMetrics)
async def get_overview(_admin: AdminUser, db: DBSession) -> OverviewMetrics:
    """Top-line platform counts — students, instructors, courses, enrollments."""
    return await metrics_service.overview(db)


@router.get("/enrollments-over-time", response_model=EnrollmentsTimeSeries)
async def get_enrollments_over_time(
    _admin: AdminUser,
    days: int = Query(30, ge=1, le=365),
) -> EnrollmentsTimeSeries:
    """Daily new-enrollment series from the Mongo event log."""
    return await metrics_service.enrollments_time_series(days=days)


@router.get("/popular-courses", response_model=PopularCourses)
async def get_popular_courses(
    _admin: AdminUser,
    db: DBSession,
    limit: int = Query(10, ge=1, le=50),
) -> PopularCourses:
    """Courses with the most enrollments."""
    return await metrics_service.popular_courses(db, limit=limit)


@router.get("/completion", response_model=CompletionMetrics)
async def get_completion(_admin: AdminUser, db: DBSession) -> CompletionMetrics:
    """Course completion rate + average time-to-complete (excludes dropped)."""
    return await metrics_service.completion(db)


@router.get("/recent-activity", response_model=RecentActivity)
async def get_recent_activity(
    _admin: AdminUser,
    limit: int = Query(20, ge=1, le=100),
) -> RecentActivity:
    """Latest events of any type, from the Mongo event log."""
    return await metrics_service.recent_activity(limit=limit)

File: frontend/src/api/metrics.ts
import { api } from "./client";
import type {
  CompletionMetrics,
  EnrollmentsTimeSeries,
  OverviewMetrics,
  PopularCourses,
  RecentActivity,
} from "./types";

export async function getOverview(): Promise<OverviewMetrics> {
  const { data } = await api.get<OverviewMetrics>("/admin/metrics/overview");
  return data;
}

export async function getCompletion(): Promise<CompletionMetrics> {
  const { data } = await api.get<CompletionMetrics>("/admin/metrics/completion");
  return data;
}

export async function getPopularCourses(limit = 10): Promise<PopularCourses> {
  const { data } = await api.get<PopularCourses>("/admin/metrics/popular-courses", {
    params: { limit },
  });
  return data;
}

export async function getEnrollmentsOverTime(days = 30): Promise<EnrollmentsTimeSeries> {
  const { data } = await api.get<EnrollmentsTimeSeries>(
    "/admin/metrics/enrollments-over-time",
    { params: { days } },
  );
  return data;
}

export async function getRecentActivity(limit = 20): Promise<RecentActivity> {
  const { data } = await api.get<RecentActivity>("/admin/metrics/recent-activity", {
    params: { limit },
  });
  return data;
}
